src/NonfungiblePositionManager.sol

```diff
@@ -0,0 +1,226 @@
+import {Multicall} from "./base/Multicall.sol";
+import {PoolInitializer} from "./base/PoolInitializer.sol";
+import {CurrencySettleTake} from "./libraries/CurrencySettleTake.sol";
+import {LiquidityRange, LiquidityRangeId, LiquidityRangeIdLibrary} from "./types/LiquidityRange.sol";
+
+contract NonfungiblePositionManager is
+    INonfungiblePositionManager,
+    ERC721Permit,
+    PoolInitializer,
+    Multicall,
+    SafeCallback
+{
+    using CurrencyLibrary for Currency;
+    using CurrencySettleTake for Currency;
+    using PoolIdLibrary for PoolKey;
+    using LiquidityRangeIdLibrary for LiquidityRange;
+    using StateLibrary for IPoolManager;
+    using TransientStateLibrary for IPoolManager;
+    using SafeCast for uint256;
+
+    /// @dev The ID of the next token that will be minted. Skips 0
+    uint256 public nextTokenId = 1;
+
+    // maps the ERC721 tokenId to its Range (poolKey, tick range)
+    mapping(uint256 tokenId => LiquidityRange range) public tokenRange;
+
+    constructor(IPoolManager _manager)
+        ImmutableState(_manager)
+        ERC721Permit("Uniswap V4 Positions NFT-V1", "UNI-V4-POS", "1")
+    {}
+
+    /// @param unlockData is an encoding of actions, params, and currencies
+    /// @return returnData is the endocing of each actions return information
+    function modifyLiquidities(bytes calldata unlockData, uint256 deadline)
+        external
+        checkDeadline(deadline)
+        returns (bytes[] memory)
+    {
+        // TODO: Edit the encoding/decoding.
+        return abi.decode(manager.unlock(abi.encode(unlockData, msg.sender)), (bytes[]));
+    }
+
+    function _unlockCallback(bytes calldata payload) internal override returns (bytes memory) {
+        // TODO: Fix double encode/decode
+        (bytes memory unlockData, address sender) = abi.decode(payload, (bytes, address));
+
+        (Actions[] memory actions, bytes[] memory params) = abi.decode(unlockData, (Actions[], bytes[]));
+
+        bytes[] memory returnData = _dispatch(actions, params, sender);
+
+        return abi.encode(returnData);
+    }
+
+    function _dispatch(Actions[] memory actions, bytes[] memory params, address sender)
+        internal
+        returns (bytes[] memory returnData)
+    {
+        if (actions.length != params.length) revert MismatchedLengths();
+        returnData = new bytes[](actions.length);
+        for (uint256 i; i < actions.length; i++) {
+            if (actions[i] == Actions.INCREASE) {
+                returnData[i] = _increase(params[i], sender);
+            } else if (actions[i] == Actions.DECREASE) {
+                returnData[i] = _decrease(params[i], sender);
+            } else if (actions[i] == Actions.MINT) {
+                // TODO: Mint will be coupled with increase.
+                returnData[i] = _mint(params[i]);
+            } else if (actions[i] == Actions.CLOSE_CURRENCY) {
+                returnData[i] = _close(params[i], sender);
```

Comment:
as an optimization, i wonder if we should add a `CLOSE_PAIR` and write an internal function which reads the deltas for the pair with 1 exttload

should help out in the vanilla cases

Reply:
I think that makes sense - what do you think @hensha256 adding this as a command in our ActionsRouter?

Reply:
honestly imo the best optimisation to make is to use `settle` and `take` instead of `close` for the vast majority of txs. The only txs that should use close are the auto compounding ones.
We can discuss this pair thing, i've been trying to think how to optimise for a regular swap too. I think it might be best t put that kind of optimisation inside the `swap` / `liquidity` commands themselves. But lets sync abut it

Reply:
for what its worth this sort of thing is why i wanted to add those `handleAdditionalCommand` functions on the actions router - so that people could make custom optimised commands that suit their usecase

Reply:
Let's sync - I've added to the list to discuss.

Reply:
https://github.com/Uniswap/v4-periphery/issues/168

Reply:
https://github.com/Uniswap/v4-periphery/issues/168